libs/nest/config/src/lib/ConfigurationLoader.ts

```diff
@@ -0,0 +1,162 @@
+      ) as Configuration<T>
+    }
+
+    try {
+      const result = this.definition.load(this)
+
+      // Loader did not crash, but there might still be issues we need to deal with.
+      return this.handleResult(result)
+    } catch (err) {
+      // Loader crashed. If there are reported issues, throw those instead.
+      return this.handleResult(undefined, err)
+    }
+  }
+
+  private handleResult(result?: T, error?: Error) {
+    const issues = Object.entries(this.issues)
+    const errorMessage = this.formatIssues(issues)
+    const hasParseIssue = issues.find(
+      (issue) => issue[1] === 'could not be parsed as JSON',
+    )
+    const hasMissingIssue = !hasParseIssue && issues.length > 0
+
+    // Always throw if there are any parse issues.
+    if (hasParseIssue) {
+      throw new ConfigurationError(errorMessage)
+    }
+
+    // Only throw missing issue in production
+    if (hasMissingIssue && !this.isDev) {
+      throw new ConfigurationError(errorMessage)
+    }
+
+    // Re-throw other unexpected errors from loader.
+    if (error) {
+      throw error
+    }
+
+    if (result === undefined) {
+      throw new ConfigurationError(
+        `Failed loading configuration for ${this.definition.name}. Config was undefined.`,
+      )
+    }
+
+    if (this.isDev && hasMissingIssue) {
+      logger.warn({
+        message: `Could not load configuration for ${this.definition.name}. Missing ${issues.length} required environment variable(s).`,
+        category: 'ConfigModule',
+      })
+      return new InvalidConfiguration(errorMessage) as Configuration<T>
+    }
+
+    this.validateSchema(result)
+
+    return {
+      ...result,
+      isConfigured: true,
+    }
+  }
+
+  private validateSchema(result: T) {
+    if (!this.definition.schema) {
+      return
+    }
+
+    const parseResult = this.definition.schema.safeParse(result)
+
+    if (!parseResult.success) {
+      throw new ConfigurationError(
+        this.formatValidationErrors(parseResult.error),
+      )
+    }
+  }
+
+  private formatIssues(issues: Array<[string, Issue]>) {
+    const formattedIssues = issues
+      .map((issue) => `- ${issue[0]} ${issue[1]}`)
+      .join('\n')
+    return `Failed loading configuration for ${this.definition.name}:\n${formattedIssues}`
+  }
+
+  private formatValidationErrors(error: ZodError) {
+    const formattedErrors = error.errors
+      .map((error) => `- ${error.path.join('.') || '<ROOT>'}: ${error.message}`)
+      .join('\n')
+    return `Failed loading configuration for ${this.definition.name}. Validation failed:\n${formattedErrors}`
+  }
+
+  required(envVariable: string, devFallback?: string): string {
+    const value = process.env[envVariable]
+    if (value !== undefined) {
+      return value
+    }
+    if (this.isDev && devFallback !== undefined) {
+      return devFallback
+    }
+    this.issues[envVariable] = 'missing'
```

Comment:
Should the known issues be a const object instead of "magic" strings?

```
const IssueReason = {
	MISSING: 'missing',
	PARSE_ERROR: 'could not be parsed as JSON',	
}
```